GeoRSS sensor

exxamalte opened the pull request:
## Description:
Generic GeoRSS events service
Retrieves current events (typically incidents or alerts) in GeoRSS format, and 
shows information on events filtered by distance to the HA instance's location 
and grouped by category.

**Related issue (if applicable):** n/a

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3124

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: geo_rss_events
    name: NSW Fire Service
    url: http://www.rfs.nsw.gov.au/feeds/majorIncidents.xml
    icon: mdi:fire
    unit_of_measurement: 'Incidents'
    categories:
      - 'Emergency Warning'
      - 'Watch and Act'
      - 'Advice'
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
generic GeoRSS service sensor

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
categories now optional; fixed longitude/latitude mix-up

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
one more example; typo corrected

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
moving from prototype to new component

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
take care of the case where no coordinates are provided

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
streamlined handling geometries

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Code style, comments, line lengths, ordering

exxamalte committed:
Merge branch 'geo-rss-sensor' of https://github.com/exxamalte/home-assistant into geo-rss-sensor

exxamalte committed:
Updated global requirements

exxamalte committed:
Updated global requirements

exxamalte committed:
Updated global requirements

exxamalte committed:
simplified and streamlined code structure

exxamalte committed:
simplified and streamlined code structure

exxamalte committed:
simplified and streamlined code structure

exxamalte committed:
simplified and streamlined code structure

exxamalte committed:
refactored code for easier testing

exxamalte committed:
refactored code for easier testing

exxamalte committed:
refactored code for easier testing

exxamalte committed:
refactored code for easier testing

exxamalte committed:
exposing guid property

exxamalte committed:
exposing guid property

exxamalte committed:
exposing guid property

exxamalte committed:
exposing guid property

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
fixes conversion of WGS84 points

exxamalte committed:
initial unit test for georss sensor component

exxamalte committed:
initial unit test for georss sensor component

exxamalte committed:
initial unit test for georss sensor component

exxamalte committed:
initial unit test for georss sensor component

exxamalte committed:
renamed sensor

exxamalte committed:
renamed sensor

exxamalte committed:
renamed sensor

exxamalte committed:
renamed sensor

exxamalte committed:
shorter default name

exxamalte committed:
shorter default name

exxamalte committed:
shorter default name

exxamalte committed:
shorter default name

exxamalte committed:
documentation bits moved to home-assistant.io repository

exxamalte committed:
documentation bits moved to home-assistant.io repository

exxamalte committed:
documentation bits moved to home-assistant.io repository

exxamalte committed:
documentation bits moved to home-assistant.io repository

exxamalte committed:
Merge branch 'geo-rss-sensor' of https://github.com/exxamalte/home-assistant into geo-rss-sensor

exxamalte committed:
Merge branch 'geo-rss-sensor' of https://github.com/exxamalte/home-assistant into geo-rss-sensor

homeassistant commented:
Hi @exxamalte, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8789).

Once you do that we will be able to review and accept this pull request.

Thanks!

exxamalte committed:
removed trailing whitespace

homeassistant commented:
Hi @exxamalte, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8789).

Once you do that we will be able to review and accept this pull request.

Thanks!

exxamalte committed:
removed over-indentation

homeassistant commented:
Hi @exxamalte, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8789).

Once you do that we will be able to review and accept this pull request.

Thanks!

exxamalte committed:
removed code snippets of unclear origin in terms of licensing and ownership

homeassistant commented:
Hi @exxamalte, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8789).

Once you do that we will be able to review and accept this pull request.

Thanks!

exxamalte committed:
satisfy lint issues

exxamalte committed:
removed ununsed import; regenerated requirements

exxamalte committed:
fix some more lint issues

exxamalte committed:
fix some more lint issues

exxamalte committed:
this should greatly improve test coverage

exxamalte committed:
fix some more lint issues

exxamalte committed:
fix some more lint issues

exxamalte commented:
I have signed the CLA a couple of times now. Is there any way to get cla-bot to recheck?

exxamalte committed:
using default update interval constant now

exxamalte committed:
better constant and variable names for scan interval

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+    scan_interval = config.get(CONF_SCAN_INTERVAL) or DEFAULT_SCAN_INTERVAL
+    unit_of_measurement = config.get(CONF_UNIT_OF_MEASUREMENT)
+
+    if None in (home_latitude, home_longitude):
```
These are actually required in the core schema, so you can skip this check:

https://github.com/home-assistant/home-assistant/blob/2dab239021cbbf265f584088f2fed06abd8c1475/homeassistant/config.py#L131-L132

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+    name = config.get(CONF_NAME)
+    icon = config.get(CONF_ICON)
+    categories = config.get(CONF_CATEGORIES)
+    scan_interval = config.get(CONF_SCAN_INTERVAL) or DEFAULT_SCAN_INTERVAL
```
Since we're using scan interval directly, my preference would be to just redefine it in the schema here, and supply the default in the schema.

https://github.com/home-assistant/home-assistant/blob/2dab239021cbbf265f584088f2fed06abd8c1475/homeassistant/helpers/config_validation.py#L402

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+        self.events = None
+
+
+class GeoRssServiceUpdater:
```
Can we merge this functionality into one class `GeoRssServiceData`? We should be able to instantiate it, create all the sensor objects, and then perform the first update.

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+        self._name = name
+        self._icon = icon
+        self._unit_of_measurement = unit_of_measurement
+        if category is not None:
```
Small style preference. Would prefer one of:
`id_base = category or 'any'`
`id_base = 'any' if category is None else category`

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+    def async_update(self, *_):  # pragma: no cover
+        """Retrieve data from GeoRSS feed and update devices."""
+        import feedparser
+        feed_data = feedparser.parse(self._url)
```
This is a blocking library. All async methods need to be converted to their synchronous equivalents.

Julius2342 replied:
you could also fetch the url asynchronously  and process the result locally with feedparser.

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+    vol.Required(CONF_URL): cv.string,
+    vol.Optional(CONF_RADIUS, default=DEFAULT_RADIUS_IN_KM): vol.Coerce(float),
+    vol.Optional(CONF_NAME, default=None): cv.string,
+    vol.Optional(CONF_ICON, default=DEFAULT_ICON): cv.icon,
```
This is can be handled in customize. Just have the entity define a default.

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,359 @@
+                if event.category == self._category:
+                    my_events.append(event)
+            self._state = my_events
+            _LOGGER.info("New state: %s", self._state)
```
Unless there's extra info to provide, this isn't necessary. Hass core will already log state changes.

emlove reviewed (changes requested):
Biggest issue is that this is using a synchronous lib. All async methods here need to be converted to their synchronous equivalents.

Other than that just a few style/small notes.

exxamalte committed:
leaner if statement

exxamalte committed:
not checking for long/lat values anymore, trusting that the core provides this information

exxamalte committed:
removed unnecessary log statement showing the new state

exxamalte committed:
changed log level and output after filtering entries

exxamalte committed:
removed icon config parameter in favour of using 'customize' to define a different icon

exxamalte committed:
refactored the use of scan_interval config parameter

exxamalte commented:
@armills: thanks for the review; I just pushed the first batch of amendments and continue with the remaining changes.

Julius2342 commented:
Looking forward to this feature!

In the documentation we should add some examples of GeoRSS feeds useful for Home Assistant, e.g. https://maps.dwd.de/geoserver/dwd/wms?service=WMS&version=1.1.0&request=GetMap&layers=dwd:Warnungen_Gemeinden&styles=&bbox=5.876914,47.270362,15.037507,55.044381&width=512&height=434&srs=EPSG:4326&format=application%2Frss%2Bxml (Official warn feed of the authorities of Germany. )

exxamalte committed:
fixed tests after refactoring code

exxamalte committed:
refactored code to use synchronous calls, and merged the updater and data class into one

exxamalte commented:
@Julius2342: Good idea. I just tried that URL and while it works in general with this new sensor, it appears that each entry contains very detailed information. Inside the standard `description` tag the details are further encoded in a proprietary format.
I would imagine that version 2 of this sensor could come with a configurable data extractor, paired with a better UI info dialogue to show those extracted details.

exxamalte committed:
empty commit to rerun ci that previously timed out

exxamalte commented:
@armills: I fixed up all the requested changes; cla-bot still not happy though. Could you please have another look at the code if the changes are OK? Thanks.

armills commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,325 @@
+            device = GeoRssServiceSensor(hass, category, data, name,
+                                         unit_of_measurement)
+            devices.append(device)
+    add_devices(devices)
```
Change this to `add_devices(devices, True)`, which will call `self.update()` before adding the devices to hass. You can then drop `self.update()` from the constructor. This is just our preferred style to avoid blocking IO in the object constructors.

exxamalte replied:
I have made the corresponding change. However, the current construct of sensor and data classes appears to require the very first update to be made outside the async routine; hence I call update on the data object before instantiating the devices.
If this update did not happen, then each sensor would call update on the shared data object independently, but most or none would wait for the result, i.e. the first sensor update yields in empty states for most or all sensors. I have seen at least the 'glances' component using the same approach.

emlove reviewed (commented):
One last small style request, otherwise this looks great! I'll grab someone to take a look at cla-bot.

exxamalte committed:
update shared data object in favour of the sensor in its constructor

exxamalte committed:
fixed tests

emlove reviewed (approved):
LGTM.

I think at this point the best thing to do is just open a new PR from this branch to get around the cla-bot issue.

pvizeli commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,328 @@
+DEFAULT_UNIT_OF_MEASUREMENT = 'Events'
+
+DOMAIN = 'geo_rss_events'
+ENTITY_ID_FORMAT = 'sensor.' + DOMAIN + '_{}'
```
Please import that from sensor.

pvizeli commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,328 @@
+                          all_events)
+            self._state = all_events
+        else:
+            # Group events by category.
```
hold that small: `self._state = [event for Event in list if Event.xy == xy]`

pvizeli commented on homeassistant/components/sensor/geo_rss_events.py:
```diff
@@ -0,0 +1,328 @@
+        return events
+
+    @staticmethod
+    def create_event(feature, distance, geometry):
```
Remove that and the Event class and hold it small.

exxamalte replied:
@pvizeli: True, the sensor's state is just the number of events found in the RSS feed filtered by distance. However, as you can see in the `device_state_attributes` method, I am displaying the title and distance in the info dialogue.
My current approach was to simply extract all information from the RSS feed that appeared useful. In a next iteration of this GeoRSS sensor, I was planning to also amend the UI and for example show a map with pins, or offer the option to order the events found by distance or by date.
Let me have a think about how to reduce the code without losing crucial information.

pvizeli reviewed (changes requested):
You should be able to handle all that with 50% less code. Don't create a event object or some other things. Your sensor is a counter, so return only the count of events.

You can fire a "Event" on event bus and make a small que for store the already sending Events. But don't store that in a object struct like now.

exxamalte committed:
removed own entity id format and import from sensor instead

exxamalte committed:
minimising code

pvizeli commented:
You can make a component with a webview and a card on webui to show rss. But as sensor let's Show the Count and maybe attributes with more Infos and fire the Events as Events to device bus. So you can make a Automation and Show it on UI with persistant_notification.

exxamalte committed:
first step in simplyfying code

exxamalte committed:
refactored how parsed georss feed's data is temporarily stored; simplified code and fixed tests

exxamalte committed:
removed unnecessary imports

exxamalte committed:
following lint's recommendation for indentation

exxamalte committed:
hound contradicted lint - new try on indentation

exxamalte commented:
@pvizeli: In the meantime I have refactored the code quite a bit - I removed the Event class in favour of a more light-weight approach to store only the required data in memory. The sensor's state is now only the number of nearby events found in the RSS feed, and the attributes are computed at the time of the update.
For now I am not using the HA event bus to populate data, but I can see the potential of adding this later, once I get more feedback from users.
What do you think now?

exxamalte committed:
Merge branch 'dev' into geo-rss-sensor

exxamalte commented:
@pvizeli, @armills: Please let me know if any more changes to the code are required for this PR. If not, I would create a new PR off the back of this - as suggested - to circumvent the issue with cla-bot. Thanks.

exxamalte commented:
Closing this pull request on favour of creating a new one for the same branch to circumvent the issue with cla-bot.
